Capacity Reservation Proposal for ID & Status (#523)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/columns/capacityreservationstatus.md

# Capacity Reservation Status

> Capacity Reservation Status indicates whether the charge represents either the consumption of the [*capacity reservation*](#glossary:capacity-reservation) identified in the CapacityReservationId column or when a *capacity reservation* is unused.

Reviewer: ```suggestion
Capacity Reservation Status indicates whether the charge represents the consumption of the [*capacity reservation*](#glossary:capacity-reservation) identified in the [CapacityReservationId](#capacityreservationid) column or the unused portion of that *capacity reservation*.
```

Author: This change seems unnecessary; is the current wording confusing?

Reviewer 2: I'm okay with current - Irena can you explain why you want this?

Reviewer: "_...when **a capacity reservation** is unused..._" caught my eye because we are not referring to the status of just any capacity reservation but specifically the one listed in the CapacityReservationId column. Therefore, we should **switch to "the"** .

I feel like the second version reads better, but if you prefer the first one, that's no problem."

> The CapacityReservationStatus column MUST be present in the billing data when the provider supports *capacity-reservations*. This column MUST be of type String, MUST be null when [CapacityReservationId](#capacityreservationid) is null, and MUST NOT be null when CapacityReservationId is not null and [Charge Category](#chargecategory) is "Usage". The CapacityReservationStatus MUST be one of the allowed values. The CapacityReservationStatus MUST label all unused *capacity reservation* charges and MUST label used *capacity reservation* charges if the provider supports it.

Reviewer: ```suggestion
The CapacityReservationStatus column adheres to the following requirements:

 * CapacityReservationStatus MUST be present in the billing data when the provider supports *capacity reservations* and MUST be of type String.
 * CapacityReservationStatus MUST be null when CapacityReservationId is null and MUST NOT be null when CapacityReservationId is not null.
 * CapacityReservationStatus MUST be one of the allowed values.
```

Notes:

- The nullability of this column tied to CapacityReservationId, so any nullability-related updates (e.g, restriction to ChargeCategory Usage, etc.) performed on CapacityReservationId will be automatically reflected.
- Not sure if we should keep the labeling-related requirement, switch to explicitly specifying when a particular value MUST be used (see PricingCategory), or leave it as it is.

Author: Im fine with the formatting changes; Im not ready to commit to removing the labeling requirement.

Reviewer 2: I'm okay with the labeling requirement as long as it doesn't require labeling used

Reviewer: Attaching a subset of Pricing Category requirements as an example. 

> * PricingCategory MUST be one of the allowed values.
> * PricingCategory MUST be "Standard" when pricing is predetermined at the agreed upon rate for the [billing account](#glossary:billing-account).
> * PricingCategory MUST be "Committed" when [CommitmentDiscountId](#commitmentdiscountid) is not null.

I prefer this wording because it’s more precise and, after all, for consistency with the rest of the spec.

Reviewer: see comment https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/523#discussion_r1738494230 for new suggestion with additional requirements

> The CapacityReservationStatus column MUST be present in the billing data when the provider supports *capacity-reservations*. This column MUST be of type String, MUST be null when [CapacityReservationId](#capacityreservationid) is null, and MUST NOT be null when CapacityReservationId is not null and [Charge Category](#chargecategory) is "Usage". The CapacityReservationStatus MUST be one of the allowed values. The CapacityReservationStatus MUST label all unused *capacity reservation* charges and MUST label used *capacity reservation* charges if the provider supports it.

Reviewer: Attaching a subset of Pricing Category requirements as an example. 

> * PricingCategory MUST be one of the allowed values.
> * PricingCategory MUST be "Standard" when pricing is predetermined at the agreed upon rate for the [billing account](#glossary:billing-account).
> * PricingCategory MUST be "Committed" when [CommitmentDiscountId](#commitmentdiscountid) is not null.

I prefer this wording because it’s more precise and, after all, for consistency with the rest of the spec.

Based on that, here is a new suggestion

```suggestion
The CapacityReservationStatus column adheres to the following requirements:

 * CapacityReservationStatus MUST be present in the billing data when the provider supports *capacity reservations* and MUST be of type String.
 * CapacityReservationStatus MUST be null when CapacityReservationId is null and MUST NOT be null when CapacityReservationId is not null.
 * CapacityReservationStatus MUST be one of the allowed values.
* CapacityReservationStatus MUST be "Used" when the charge represents consumption of the *capacity reservation*.
* CapacityReservationStatus MUST be "Unused" when the charge represents the unused portion of the *capacity reservation*.
```

# Capacity Reservation Status

Capacity Reservation Status indicates whether the charge represents either the consumption of the [*capacity reservation*](#glossary:capacity-reservation) identified in the CapacityReservationId column or when the *capacity reservation* is unused.

The CapacityReservationStatus column adheres to the following requirements:
 * CapacityReservationStatus MUST be present in the billing data when the provider supports *capacity reservations* and MUST be of type String.

>  * CapacityReservationStatus MUST be null when CapacityReservationId is null and MUST NOT be null when CapacityReservationId is not null.

Reviewer 2: SHOULD NOT be null when CapacityReservationId is not null and the charge is related to a Used or Unused capacity reservation.

>  * CapacityReservationStatus MUST be null when CapacityReservationId is null and MUST NOT be null when CapacityReservationId is not null.

Reviewer 2: ```suggestion
 * CapacityReservationStatus MUST be null when CapacityReservationId is null and SHOULD NOT be null when CapacityReservationId is not null and a capacity reservation status can be applied.
```

```

>  * CapacityReservationStatus MUST be null when CapacityReservationId is null and SHOULD NOT be null when CapacityReservationId is not null.

Reviewer: ```suggestion
 * CapacityReservationStatus MUST be null when CapacityReservationId is null 
 * CapacityReservationStatus MUST NOT be null when CapacityReservationId is not null and [Charge Category](#chargecategory) is "Usage".
```

This is how CommitmentDiscountStatus requirement is specified

> MUST be null when [CommitmentDiscountId](#commitmentdiscountid) is null, and MUST NOT be null when CommitmentDiscountId is not null and [Charge Category](#chargecategory) is "Usage".

Considering, we wish to allow nulls for Purchase records, this should work CapacityReservationStatus, right?

Reviewer 2: @gparker-at-sf I approve Irena's suggestion. 

I wish I wouldve thought of it earlier because it actually is consistent with Commitment purchases where we don't require labeling used vs. unused when you purchase a SP or RI. This is effectively the same scenario but the "purchase" of the ODCR is 0 cost.

# Capacity Reservation Status

Capacity Reservation Status indicates whether the charge represents either the consumption of the [*capacity reservation*](#glossary:capacity-reservation) identified in the CapacityReservationId column or when the *capacity reservation* is unused.

The CapacityReservationStatus column adheres to the following requirements:

>  * CapacityReservationStatus MUST be present in the billing data when the provider supports *capacity reservations* and MUST be of type String.

Reviewer 3: ```suggestion
 * CapacityReservationStatus MUST be present in a FOCUS dataset when the provider supports *capacity reservations* and MUST be of type String.
```

 * CapacityReservationStatus MUST be null when CapacityReservationId is null.
 * CapacityReservationStatus MUST NOT be null when CapacityReservationId is not null and [ChargeCategory](#chargecategory) is "Usage".
 * CapacityReservationStatus MUST be one of the allowed values.

>  * CapacityReservationStatus MUST label all unused *capacity reservation* charges and MUST label used *capacity reservation* charges if the provider supports it.

Reviewer 3: Is this saying that all rows need to have a CapacityReserationStatus? If so, that's already covered by the other requirements. If not, I'm confused by what this is saying. Given the requirements are already stated above and we don't have (or need) this for CommitmentDiscountStatus, I would propose we remove it.
```suggestion
```

Reviewer 2: I think line 10 here just indicates labeling unused ODCR is always required, but labeling used ODCR is only required if supported by provider.
